### ISSUE TITLE
embed tinygo so that golang can be compiled on the fly to wasm.

### ISSUE THREAD

**Issue by gedw99** [tetratelabs/wazero#807]
**Is your feature request related to a problem? Please describe.**

User Defined Function ( UDF ) as where the user can create functions, and so requires that the golang code be compiled to wasm on the fly without recompile.

**Describe the solution you'd like**

I provide a Web area to edit golang, and wazero recompiles the wasm at runtime.

**Describe alternatives you've considered**

**Additional context**

Am trying to think of ways to do this.

Embed tinygo and then extract it to the file system at init time. 
Run with docker compose with one docker for wazero and one for tinygo.

**Comment by codefromthecrypt** [comment]
@gedw99 Thanks for raising this. It isn't entirely likely we'll be able to embed tinygo easily as it doesn't have a mechanism to [embed as a library](https://github.com/tinygo-org/tinygo/issues/2016). You'll notice on that issue there are some substantial system dependencies, and that it forks other OS processes as well.

That said, we've been repeatedly asked for something like this, as it is intuitive. Ex. folks assume an embeddable Go runtime would have some relationship with an embeddable Go compiler. The reason is that while some people look at wasm as polyglot, others look at it as a way to extend Go with Go safely. I think we can keep this open at least until the end of the year to figure out if there are any ways to solve Go->wasm without forking. I'll re-word the desc a bit to help inventory thoughts.

Anyone who wants something like this to happen should 👍 the issue and also describe what they are trying to achieve. cc @simar7 @james-lawrence

PS a docker orchestration way sounds like the "today" way, and seems like someone could make a proof-of-concept for that. Particularly re-using tinygo's docker image will contain the deps to just docker and go using it.

**Comment by codefromthecrypt** [comment]
Adjusted the desc @gedw99 feel free to revise again if I got you wrong. I will add any other use cases and ideas as they collect here. Also, if I find any github repos doing any similar pattern I'll also revise. Thanks again!

**Comment by gedw99** [comment]
I am thinking of running the wasm store and compiler in sone docker thing.

which means I need a registry of sorts that knows where the wasm is stored etc . 

i think spin has this.

capsule which uses Warri also has a registry now too.

**Comment by codefromthecrypt** [comment]
As you rightly mention, there are quite a few tools out there that have solutions for wasm distribution. Let's assume that wasm distribution is mostly solved (ex via file systems, OCI repositories or plain HTTP usually)

An embedded Go compiler if embedded in the same process may never have to distribute wasm. If distributed, indeed it can likely use one of the practices in use, as that's a good way to decouple producer and consumer of wasm!

**Comment by gedw99** [comment]
I agree that docker is the way to get it done today. I wonder what the APi would look like. 

I prefer send the code and get back the wasm file as bytes to make is simple. 
Wonder what options need to be added that Wazero needs when telling tinygo what to do. For example a debug or release build.

**Comment by lthibault** [comment]
Naive question:  is it possible to compile TinyGo itself into WASM and execute it in Wazero, providing host functions in lieu of syscalls?

**Comment by codefromthecrypt** [comment]
@lthibault yes, typically people will use tinygo's [wasi target](http://wazero.io/languages/tinygo/) as it is less work. Here's an [example](https://github.com/tetratelabs/wazero/tree/main/imports/wasi_snapshot_preview1/example).

**Comment by gedw99** [comment]
> As you rightly mention, there are quite a few tools out there that have solutions for wasm distribution. Let's assume that wasm distribution is mostly solved (ex via file systems, OCI repositories or plain HTTP usually)
> 
> An embedded Go compiler if embedded in the same process may never have to distribute wasm. If distributed, indeed it can likely use one of the practices in use, as that's a good way to decouple producer and consumer of wasm!

Great points !! 

I can see the following ways:

1. embedded; embed the tinygo compiler into some golang app using the standard golang embed annotation.  
2. Non embedded: Curl from ( https://github.com/tinygo-org/tinygo/releases) at runtime, and then just call it. 
3. decoupled: Docker of tinygo running somewhere.
4. Probably many others ....

All  of the above need an API ( local and networked ) using GRPC ( for example ) that exports the various functions that developers need. A CTL / CLI using GRPC would likely be the mechanism. 
 
What types of things would a dev need from this API ?
1. compiling and the options associated with it. Source being local or remote. Result being binary stream or S3 URI.
2. Inspecting to check the WASI exports.
3. Debugging the WASI. Thats obviously a huge one.
4. Probably many more i have not thought of.

My use case is a Low code style platform where Functions are written within the Platform GUI at "Runtime". These platforms blue the distinction between "DesignTime", "CompileTIme", "DeplyoyTime", "OpsTime".

**Comment by codefromthecrypt** [comment]
thanks for starting the gears turning.

So, on 1 or 4) one issue is how to handle non-standard libraries (ex go.mod etc). So, one approach could be to forbid and another might be to supply a git URL instead of a source file.

In general, I think an issue like this would have to keep to the minimum requirements, not maximum. I wouldn't expect this issue to get into things like S3 or other cloud storage, rather a higher level abstraction like OCI or http if anything. Similarly, since this is an issue on the wazero repo and we're already pretty far out of scope, things like runtime debugging could be a bridge too far.

Usually issues have a result of a change on the repo in other words, and I'm thinking this is already a bit beyond that. Until we find a better venue, though.. carry on :D
